File: LeeC/309_BestTimeToBuyAndSellStockWithCooldown/Solution.java



class Solution {
    
    public int maxProfit(int[] prices) {
   	for(int i=0; i<prices.length; i++)
	    System.out.print(prices[i] + "  ");
	System.out.println();
	 
        int sell = 0, prev_sell = 0, buy = Integer.MIN_VALUE, prev_buy;
        for (int price : prices) {
            prev_buy = buy;
            buy = Math.max(prev_sell - price, prev_buy);
            prev_sell = sell;
            sell = Math.max(prev_buy + price, prev_sell);
        }

	System.out.println("sell = " + sell);
        return sell;
    }


    public static void main(String args[]){
	Solution s1 = new Solution();

	int[] prices = new int[]{1, 2, 3, 0, 2};
	s1.maxProfit(prices);

    }    
}

File: LeeC/309_BestTimeToBuyAndSellStockWithCooldown/Solution.py





class Solution(object):
    def maxProfit(self, prices):
	for p in prices:
	    print p,
	print 

        if len(prices) < 2:
            return 0
        sell, buy, prev_sell, prev_buy = 0, -prices[0], 0, 0
        for price in prices:
            prev_buy = buy
            buy = max(prev_sell - price, prev_buy)
            prev_sell = sell
            sell = max(prev_buy + price, prev_sell)
        print "sell = ", sell, "\n"
	return sell
        

s1 = Solution()

prices = [1, 2, 3, 0, 2]
s1.maxProfit(prices)
